packages/tools/src/tools/FreehandROISculptorTool.ts

```diff
@@ -0,0 +1,685 @@
+   *
+   * @param pushedHandles - The first and last handles that were pushed.
+   */
+  _findNewHandleIndices(pushedHandles: {
+    first: number | undefined;
+    last: number | undefined;
+  }): Array<number> {
+    const { points, maxSpacing } = this._sculptData;
+    const indicesToInsertAfter = [];
+
+    for (let i = pushedHandles.first; i <= pushedHandles.last; i++) {
+      this._checkSpacing(i, points, indicesToInsertAfter, maxSpacing);
+    }
+
+    return indicesToInsertAfter;
+  }
+
+  /**
+   * Inserts a handle on the surface of the circle defined by toolSize and the mousePoint.
+   *
+   * @param insertIndex - The index to insert the new handle.
+   */
+  _insertHandleRadially(insertIndex: number): void {
+    const { points } = this._sculptData;
+
+    if (
+      insertIndex > points.length - 1 &&
+      this._commonData.isEditingOpenContour
+    ) {
+      return;
+    }
+
+    const previousIndex = insertIndex - 1;
+    const nextIndex = this._getNextHandleIndexBeforeInsert(
+      insertIndex,
+      points.length
+    );
+    const insertPosition = this._getInsertPosition(previousIndex, nextIndex);
+    const handleData = insertPosition;
+
+    points.splice(insertIndex, 0, handleData);
+  }
+
+  /**
+   * Calculates the position that a new handle should be inserted.
+   *
+   * @param previousIndex - The previous index
+   * @param nextIndex - The next index
+   */
+  _getInsertPosition(previousIndex: number, nextIndex: number): Types.Point3 {
+    let insertPosition: Types.Point2;
+    const { points, toolSize, element, mouseCanvasPoint } = this._sculptData;
+    const enabledElement = getEnabledElement(element);
+    const { viewport } = enabledElement;
+    const previousCanvasPoint = viewport.worldToCanvas(points[previousIndex]);
+    const nextCanvasPoint = viewport.worldToCanvas(points[nextIndex]);
+
+    const midPoint: Types.Point2 = [
+      (previousCanvasPoint[0] + nextCanvasPoint[0]) / 2.0,
+      (previousCanvasPoint[1] + nextCanvasPoint[1]) / 2.0,
+    ];
+
+    const distanceToMidPoint = point.distanceToPoint(
+      mouseCanvasPoint,
+      midPoint
+    );
+
+    if (distanceToMidPoint < toolSize) {
+      const directionUnitVector = {
+        x: (midPoint[0] - mouseCanvasPoint[0]) / distanceToMidPoint,
+        y: (midPoint[1] - mouseCanvasPoint[1]) / distanceToMidPoint,
+      };
+
+      insertPosition = [
+        mouseCanvasPoint[0] + toolSize * directionUnitVector.x,
+        mouseCanvasPoint[1] + toolSize * directionUnitVector.y,
+      ];
+    } else {
+      insertPosition = midPoint;
+    }
+
+    const worldPosition = viewport.canvasToWorld(insertPosition);
+
+    return worldPosition;
+  }
+
+  /**
+   * Pushes the points radially away from the mouse if they are
+   * contained within the circle defined by the freehandSculpter's toolSize and
+   * the mouse position.
+   *
+   * @param viewport
+   */
+  _pushHandles(viewport: Types.IViewport) {
```

Comment:
Again, can you make this protected pushHandles so it could be extended or re-used?

Reply:
Updated